[documentation]: Review ISI and WASM topics (#64)
hyperledger/iroha-2-docs: src/guide/advanced/wasm.md

- The ISI syntax is verbose and ugly.
- The ISI syntax is not familiar for most programmers.

> - While ISI smart contracts are compact (usually a few bytes), they need to

Reviewer: ```suggestion
- While simple ISI smart contracts are compact (usually a few bytes), they need different kinds of 
```

>   be manually optimised.

Author: ```suggestion
  manual optimisation.
```

traditional `if` statements and `for` loops.

However, in the interim, we have decided to use another portable binary
standard called **Web assembly**, or **WASM**.

## Working with WASM

The main advantage of using the WASM format is that you can use any
language you like (as long as it links statically against our helper
library), and produce a 32-bit portable executable. The compilers take care
of the optimisation, and you don't have to learn a new language (ahem...
solidity... ahem), just to operate on the blockchain.

You'd still need to use ISI from inside your WASM binary to do anything

> on-chain, but you can do other things, and write in Rust instead of

Reviewer: ```suggestion
on-chain, as we explained earlier. While the language of ISI is turing-complete (and in theory that means that you can do anything you want), it's less convenient (and efficient) to use metadata as memory, and write complex conditionals using just the tools that we've provided in the `Expression` and ISI infrastructure. Simply put it's easier to write in a well-known programming language (Rust) and build the logic out of simple instructions, rather than trying to reinvent the wheel using ISI. 
```

> composing Generic `structs` inside a macro to get conditional logic.

Author: Not sure about the second part of the sentence starting with "but". Maybe I'm missing something but it doesn't connect with the first half of the sentence

Reviewer: Added an explanation.

Moreover, as long as you fit within the limits of WASM runtime and the
provided libraries, you can do anything (and everything) you want. The
drawback is that this process is a tad more involved than just writing the
ISI using the client libraries.

> ## Simple Rust Smart Contracts

Reviewer: ```suggestion
## Simple Rust Smart Contract Example
```

WASM projects, just like any other binary in Rust, need to be separate
crates. Don't worry, it doesn't have to be big.

### 1. Create a new project

To get started, create a new project:

```bash
cargo new --lib
```

> Yes, we need the `lib` type, more on that later.

Reviewer: ```suggestion
Yes! We need the `lib` type; more on that later.
```
